server/src/main/java/org/opensearch/index/shard/RefreshListeners.java

```diff
@@ -256,51 +334,103 @@ public void afterRefresh(boolean didRefresh) throws IOException {
         /* Grab the current refresh listeners and replace them with null while synchronized. Any listeners that come in after this won't be
          * in the list we iterate over and very likely won't be candidates for refresh anyway because we've already moved the
          * lastRefreshedLocation. */
-        List<Tuple<Translog.Location, Consumer<Boolean>>> candidates;
-        synchronized (this) {
-            candidates = refreshListeners;
-            // No listeners to check so just bail early
-            if (candidates == null) {
-                return;
+        if (indexShard != null && indexShard.indexSettings.isSegRepEnabled() && indexShard.routingEntry().primary() == false) {
```

Comment:
We shouldn't need this condition either, we can see if there are any registered seqNo listeners?